src/core/include/openvino/core/type.hpp

```diff
@@ -78,37 +78,55 @@ OPENVINO_API
 std::ostream& operator<<(std::ostream& s, const DiscreteTypeInfo& info);
 
 #if defined(__ANDROID__) || defined(ANDROID)
-#    define OPENVINO_DYNAMIC_CAST
+inline constexpr auto use_ov_dynamic_cast = true;
+#else
+inline constexpr auto use_ov_dynamic_cast = false;
 #endif
 
-/// \brief Tests if value is a pointer/shared_ptr that can be statically cast to a
-/// Type*/shared_ptr<Type>
-template <typename Type, typename Value>
-typename std::enable_if<
-    std::is_convertible<decltype(std::declval<Value>()->get_type_info().is_castable(Type::get_type_info_static())),
-                        bool>::value,
-    bool>::type
-is_type(Value value) {
-    return value && value->get_type_info().is_castable(Type::get_type_info_static());
+/// \brief Tests whether pointer can be static casted to a To*/shared_ptr<To>
+template <typename To, typename From>
+typename std::enable_if_t<
+    std::is_convertible_v<decltype(std::declval<From>().get_type_info().is_castable(To::get_type_info_static())), bool>,
+    bool>
+is_type(From* ptr) {
+    if constexpr (use_ov_dynamic_cast) {
+        return ptr && ptr->get_type_info().is_castable(To::get_type_info_static());
+    } else {
+        return dynamic_cast<std::remove_cv_t<To>*>(const_cast<std::remove_cv_t<From>*>(ptr)) != nullptr;
```

Comment:
Why `remove_cv_t`  has been added?

`is_type` should return false if `From` is `const Type` and `To` is not const. Const type cannot be casted to non-const.
Remove `const_cast`.

It looks like changed functionality, can some unit test be added/updated.

Reply:
I'd agree if it didn't change current behavior of `is_type<>()` which just compares `DiscreteTypeInfo` objects and disregards From To qualifiers. Note that `is_type` can be and is used stand alone (apart from `as_type`) in non casting flow in which case cv qualifiers don't matter. Enforcing qualifiers equality we risk breaking logic in hard-to-grep code or user application.

Reply:
Ok for `std::remove_cv_t`, as `is_type` currently ignores `constness`, but `const_cast` should be removed, use type traits to get the type or add const if required.

Reply:
Done.